### PR TITLE
Add vm to authz test

### PR DESCRIPTION
**Pull request by JimmyCYJ**
This splits out of #31618. 
Add VM to TestAuthorization_Conditions, TestAuthorization_GRPC, TestAuthorization_Path and TestAuthorization_Audit.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: tests/integration/security/authorization_test.go [hunk 1]
```diff
@@ -940,37 +940,35 @@ func TestAuthorization_Conditions(t *testing.T) {
...
-						"NamespaceC": nsC.Name(),
-						"IpA":        getWorkload(podAWithIPA, t).Address(),
-						"IpB":        getWorkload(podBWithIPB, t).Address(),
-						"IpC":        IPC,
-						"PortC":      fmt.Sprintf("%d", portC),
-					}
-
-					policies := tmpl.EvaluateAllOrFail(t, args, file.AsStringOrFail(t, "testdata/authz/v1beta1-conditions.yaml.tmpl"))
-					t.Config().ApplyYAMLOrFail(t, "", policies...)
+			c := apps.C.Match(echo.Namespace(nsC.Name()))
```
Comment (stevenctl) [inline tests/integration/security/authorization_test.go:943]: I'm a bit confused why the security tests use `echo.Namespace` matcher. Looking at the set of apps, it seems we don't have instances where `apps.{A, B, C, etc.}` contain instances from multiple namespaces.

File: tests/integration/security/authorization_test.go [hunk 2]
```diff
@@ -940,37 +940,35 @@ func TestAuthorization_Conditions(t *testing.T) {
...
+			vm := apps.VM.Match(echo.Namespace(nsA.Name()))
+			for _, cSet := range []echo.Instances{c, vm} {
+				for _, a := range apps.A.Match(echo.Namespace(nsA.Name())) {
+					a, b := a, apps.B.Match(echo.InCluster(a.Config().Cluster)).Match(echo.Namespace(nsB.Name()))
```
Comment (stevenctl) [inline tests/integration/security/authorization_test.go:947]: can we do something where we assert `apps.B.Match` has length 1, then take `[0]` here rather than all the `[0]`'s in `newTestCase` calls?

File: tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl [hunk 3]
```diff
@@ -54,7 +54,7 @@ metadata:
 spec:
   selector:
     matchLabels:
-      app: c
+      app: {{ .cSet }}
```
Comment (stevenctl) [inline tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl:57]: I don't really see any advantage to templating here, if `cSet` is always `c`? Unless the meaning of `cSet` here becomes "third deployment". 

Also what does `set` mean here? Seems like it's more like `cSvc`

File: tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl [hunk 4]
```diff
@@ -96,14 +96,14 @@ metadata:
...
   rules:
   - to:
     - operation:
-        paths: ["/source-ip-notValues-b"]
+        paths: ["/source-ip-notValues-{{ .b }}"]
     when:
     - key: source.ip
-      notValues: ["{{ .IpB }}"]
+      notValues: ["{{ .ipB }}"]
```
Comment (stevenctl) [inline tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl:106]: Pretty much the same comment as above for all of the template vars. 

It's probably outside the scope of this PR, but I think adding templating now just increases the complexity without adding value. 

Eventually, we should probably look at something like the networking tests have, where the config lives on the test case:

https://github.com/istio/istio/blob/da6178604559bdf2c707a57f452d16bee0de90c8/tests/integration/pilot/common/traffic.go#L44-L47

And use a somewhat standard set of template vars and allowing custom template vars for individual tests that need them:

https://github.com/istio/istio/blob/da6178604559bdf2c707a57f452d16bee0de90c8/tests/integration/pilot/common/traffic.go#L93-L105


Ideally, we merge the security context stuff, the traffictestcase and rbacutil, but also seems like a bigger effort. 


In short – it's probably better not to add templating now, it seems like it would add more work to refactor and use `echotest`. I don't feel strongly though.

**Review by stevenctl — APPROVED** [review]
Most of the comments are nits, so approving but adding a hold.

File: tests/integration/security/authorization_test.go (hunk 1 above)
Comment (JimmyCYJ) [reply tests/integration/security/authorization_test.go:943]: By refactoring tests in authorization_test.go, my understanding is, there are some security policies that allows/denies requests by checking sender's namespace. E.g. a workload `d` allows request sent by `c` from ns1, and denies request sent by `c` from ns2. There are also some security policies that allows/denies request by checking principals. E.g. a workload `c` from ns1 has principal a `cluster.local/ns/ns1/sa/c`, if workload `d` only allows request with this principal, then `d` will deny requests by `c` from other namespaces. There are other tests that verify the security policy works correctly per workload.

File: tests/integration/security/authorization_test.go (hunk 2 above)
Comment (JimmyCYJ) [reply tests/integration/security/authorization_test.go]: Good point. Updated. Thanks!

File: tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl (hunk 3 above)
Comment (JimmyCYJ) [reply tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl:57]: According to reviews on other PRs, changing the variables to src0, src1, ..., dst0, dst1,... makes it hard to review, because it is not clear whether coverage is persistent. It is suggested to keep using existing variables. Here `cSet` denotes a third deployment, which could be App C running on a pod. When we have support for multi-VM, we could pass a VM deployment to `cSet`.

File: tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl (hunk 4 above)
Comment (JimmyCYJ) [reply tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl:106]: I had discussed with @yangminzhu , this is because the security tests require multiple workloads as destination. e.g. some tests apply multiple security policies to multiple workloads, and expect each workload to behave differently. e.g. a workload `a` will receive different response code when sending requests to `b`, `c`, and `d`. There are tests which verify that different source workloads sending request to the same destination will receive different response code. 

Originally I tried to split the policy to small pieces, and apply to the single destination workload (pod/vm), the feedback from the review is that this approach reduces the test coverage. So I reverted that change back and keep using workloads `a`, `b`, `c`... etc. Once we have support for multiple VMs, I can add a group of VMs to rerun the tests in one more round (e.g. pass vm a, vm b, vm c to those `a`, `b`, `c`...).

File: tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl (hunk 3 above)
Comment (JimmyCYJ) [reply tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl:57]: https://github.com/istio/istio/pull/32016 also reverted some variable names back for readability concern.

File: tests/integration/security/authorization_test.go (hunk 1 above)
Comment (stevenctl) [reply tests/integration/security/authorization_test.go:943]: I missed before that `c` is not unique to a single namespace.

Comment (stevenctl) [reply tests/integration/security/authorization_test.go:943]: https://github.com/istio/istio/pull/32121 – this should make the tests easier to differentiate :)

File: tests/integration/security/authorization_test.go [hunk 5]
```diff
@@ -1061,61 +1062,64 @@ func TestAuthorization_GRPC(t *testing.T) {
...
-			echoboot.NewBuilder(t).
-				With(&a, util.EchoConfig("a", ns, false, nil)).
-				With(&b, util.EchoConfig("b", ns, false, nil)).
-				With(&c, util.EchoConfig("c", ns, false, nil)).
-				With(&d, util.EchoConfig("d", ns, false, nil)).
-				BuildOrFail(t)
+			ns := apps.Namespace1
+			c := apps.C.Match(echo.Namespace(apps.Namespace1.Name()))
+			d := apps.D.Match(echo.Namespace(apps.Namespace1.Name()))
+			for _, a := range []echo.Instances{apps.A.Match(echo.Namespace(apps.Namespace1.Name())), apps.VM.Match(echo.Namespace(apps.Namespace1.Name()))} {
```
Comment (yangminzhu) [inline tests/integration/security/authorization_test.go:1068]: can you put `[]echo.Instances{apps.A.Match(echo.Namespace(apps.Namespace1.Name())), apps.VM.Match(echo.Namespace(apps.Namespace1.Name()))}` in a separate variable out of the loop? Make it a bit easier to read, same below.

File: tests/integration/security/authorization_test.go [hunk 6]
```diff
@@ -1212,28 +1213,80 @@ func TestAuthorization_Audit(t *testing.T) {
...
-				newTestCase(c, "/deny", false),
-				newTestCase(d, "/audit", true),
-				newTestCase(d, "/other", true),
-			}
 
-			args := map[string]string{
-				"Namespace":     ns.Name(),
-				"RootNamespace": istio.GetOrFail(t, t).Settings().SystemNamespace,
-			}
+			for _, a := range []echo.Instances{apps.A.Match(echo.Namespace(ns.Name())), apps.VM.Match(echo.Namespace(ns.Name()))} {
```
Comment (yangminzhu) [inline tests/integration/security/authorization_test.go:1217]: a is the source workload sending the request right? I don't think replacing it with vm has any benefits, the destination pod should be replaced with vm as the audit log happens only on the server side.

File: tests/integration/security/authorization_test.go [hunk 7]
```diff
@@ -1212,28 +1213,80 @@ func TestAuthorization_Audit(t *testing.T) {
...
 
-			applyPolicy := func(filename string, ns namespace.Instance) {
-				policy := tmpl.EvaluateAllOrFail(t, args, file.AsStringOrFail(t, filename))
-				t.Config().ApplyYAMLOrFail(t, ns.Name(), policy...)
+						rbacUtil.RunRBACTest(t, cases)
+					})
 			}
 
-			applyPolicy("testdata/authz/v1beta1-audit.yaml.tmpl", ns)
+			// (TODO)JimmyCYJ: Support multiple VMs and apply audit policies to multiple VMs for testing.
```
Comment (yangminzhu) [inline tests/integration/security/authorization_test.go:1245]: can you also mention the following duplicate is really a short hack and should be removed (merged) with the above test cases?

File: tests/integration/security/authorization_test.go (hunk 5 above)
Comment (JimmyCYJ) [reply tests/integration/security/authorization_test.go]: Done.

File: tests/integration/security/authorization_test.go (hunk 6 above)
Comment (JimmyCYJ) [reply tests/integration/security/authorization_test.go]: I have removed the for loop, now only a is chosen as sender.

File: tests/integration/security/authorization_test.go (hunk 7 above)
Comment (JimmyCYJ) [reply tests/integration/security/authorization_test.go:1268]: Done.

**Review by yangminzhu — APPROVED** [review]
thanks!
